### ISSUE TITLE
Enhancements/suggestions

### ISSUE THREAD

**Issue by Dynious** [dmillerw/ASMTemplates#1]
This is a list of enhancements and suggestions for this library:
- [ ] Add easy getter and setter methods for fields in the 'unknown' superclass.
- [x] SubclassGenerator#generateSubclass should return a a class with the type of the template.
- [ ] ~~Maybe something to to replace references to a class? Sometimes classes create a new instance of themselves and it would be more easy if we didn't have to replace the whole method just to replace the instance creation to our own subclass.~~
- [ ] Something to easily invoke methods in the 'unknown' superclass.
- [ ] Calling super constructors from subclasses.

EDIT:
More explanation for this point:

![](http://i.imgur.com/7NukCfD.png)
It's obvious why this doesn't work, and it will probably not be possible in this way. I want to call the constructor of the 'unknown' superclass here. The constructor of my new subclass has less parameters than its superclass.
Basically this functionality:

![](http://i.imgur.com/eonUKyB.png)
- [ ] Allow Template instance creation for usage of variables in superclass construction.

This in combination with the point above could be very powerful to create a lot of classes with different parameters (some classes require constructors without parameters). Like this:

![](http://i.imgur.com/sAfbBDL.png)
- [ ] Allow subclassing of template files. (Check methods & fields in superclasses of templates)

Of course some of these things are possible with some easy Reflection, but this would just make the library more complete.

**Comment by dmillerw** [comment]
5 and 6 could prove to be tricky.

See, the template class is never actually used or called upon. It's scanned upon the generation of the subclass, but then all its fields and methods are simply _copied_ over to the super class, and any invocations are redirected.

**Comment by dmillerw** [comment]
1 and 4 are simpler than I thought. Just add the field or method you need to access, but don't annotate it. Any calls to it will just get redirected as normal to the super class

**Comment by dmillerw** [comment]
Calling super constructors is going to be a pain in the ass :(

Either you'll have to specify the signature, or shadow it. 

More fun for me!

**Comment by dmillerw** [comment]
Finally, you'll need to elaborate on 3

**Comment by Dynious** [comment]
5 is pretty much necessary to make this work right :( , 6 would be awesome, but isn't necessary (just less code writing for the end user :P).

About 3, this will probably not be possible with things breaking a lot, you can just ignore that one :P

**Comment by Dynious** [comment]
More problems :(

![](http://i.imgur.com/YnPeyuD.png)

The return value and parameter are part of a project I don't want to depend on. :S

**Comment by Dynious** [comment]
Added: Allow subclassing of template files. (Check methods & fields in superclasses of templates)
